EIPS/eip-2029.md

```diff
@@ -0,0 +1,61 @@
+discussions-to: https://ethereum-magicians.org/t/eip-2029-state-counters-contract-change-a-from-state-rent-v3-proposal/3279
+status: Draft
+type: Standards Track
+category: Core
+created: 2019-05-15
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+A smart contract is deployed on all Ethereum networks (including testnets), with the code that simply reads the slot in its storage specified by the
+only parameter. Later, this contract becomes "special" in that Ethereum start writing state counters (number of total transactions, number of accounts,
+etc.) into that contract.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+It is part of the State Rent roadmap. This particular change introduces a place in the Ethereum state where various state counters can be stored. At this
+point, the most important counter is the total number of transactions happened, and this counter will be used to populate the nonces of newly created
+non-contract accounts. This way of populating nonce ensures replay protection for accounts that were evicted and then brought back by sending ether to them.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+Ethereum currently does not have a special place in the state for tracking state counters such as number of transactions or number of accounts.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+Prior to the block A, a contract is deployed with the following code:
+`0x60 0x20 0x60 0x00 0x80 0x80 0x35 0x54 0x90 0x52 0xF3`, which corresponds to this assembly:
+`PUSH1 32 PUSH1 0 DUP1 DUP1 CALLDATALOAD SLOAD SWAP1 MSTORE RETURN`
+Call to this contract accepts one 32-byte argument, `x`, and returns the value of the storage item [`x`].
+
+This contract is deployed using `CREATE2` opcode in such a way that it has the same address on any networks.
```

Comment:
I was about to suggest this =) This does mean we have to create a signature which can be replayed, right, not an EIP 155 sig? (To my understanding EIP 155 is on wallet level not on protocol level?)

Reply:
However why not introduce this contract as a system contract at a certain address?

Reply:
I do not see why use Nick's method, because CREATE2 just makes it so much more simple (when Nick wrote it, CREATE2 did not exist). As to the "system" contract, we do not really have a notion of a "system" contract yet

Reply:
I just don't see how CREATE2 could be used in this instance though. The address for a CREATE2 depends on the address the of the creator, and CREATE2 is also an opcode so would have to be done from another contract, whose address depends on the creator's address. So to get a constant CREATE2'd contract address across chains, you'd have to use Nick's method to create the initial that creates the state counters contract. Which is more complicated than just using Nick's method.